src/toil/jobStores/abstractJobStore.py

```diff
@@ -114,10 +158,118 @@ def _checkJobStoreCreation(create, exists, jobStoreString):
+        :param urlparse.ParseResult url: The parsed url given to importFile.
+        :return: The job store file ID of the imported file
+        :rtype: str
+        """
+        with self.writeFileStream() as (writable, jobStoreFileID):
+            otherCls.readFromUrl(url, writable)
+            return jobStoreFileID
+
+    def exportFile(self, jobStoreFileID, dstUrl):
+        """
+        Exports file to destination pointed at by the destination URL.
+
+        Refer to AbstractJobStore.importFile documentation for currently supported URL schemes.
+
+        Note that the helper method _exportFile is used to read from the source and write to
+        destination. To implement any optimizations that circumvent this, the _exportFile method
+        should be overridden by subclasses of AbstractJobStore.
+
+        :param str jobStoreFileID: The id of the file in the job store that should be exported.
+        :param str dstUrl: URL that points to a file or object in the storage mechanism of a
+                supported URL scheme e.g. a blob in an Azure Blob Storage container.
+        """
+        url = urlparse.urlparse(dstUrl)
+        return self._exportFile(findJobStoreForUrl(url), jobStoreFileID, url)
+
+    def _exportFile(self, otherCls, jobStoreFileID, url):
+        """
+        Refer to exportFile docstring for information about this method.
+
+        :param toil.jobStores.abstractJobStore.AbstractJobStore otherCls: The class of job store to import from.
+        :param str jobStoreFileID: The id of the file that will be exported.
+        :param urlparse.ParseResult url: The parsed url given to importFile.
+        """
+        with self.readFileStream(jobStoreFileID) as readable:
+            otherCls.writeToUrl(readable, url)
+
+    @abstractstaticmethod
+    def readFromUrl(url, writable):
+        """
+        Reads the contents of the object at the specified location and writes it to the given
+        writable stream.
+
+        Refer to AbstractJobStore.importFile documentation for currently supported URL schemes.
+
+        :param str url: URL that points to a file or object in the storage mechanism of a
+                supported URL scheme e.g. a blob in an Azure Blob Storage container.
+        :param writable: a writable stream
+        """
+        raise NotImplementedError()
+
+    @abstractstaticmethod
+    def writeToUrl(readable, url):
+        """
+        Reads the contents of the given readable stream and writes it to the object at the
+        specified location.
+
+        Refer to AbstractJobStore.importFile documentation for currently supported URL schemes.
+
+        :param str url: URL that points to a file or object in the storage mechanism of a
+                supported URL scheme e.g. a blob in an Azure Blob Storage container.
+        :param readable: a readable stream
+        """
+        raise NotImplementedError()
+
+    @abstractclassmethod
+    def supportsUrl(cls, url):
```

Comment:
If only methods of AbstractJobStore or a subclass thereof call these methods, I think we should.